### PR TITLE
New server handler parser; Checks receiver, updates ins and outs, ignores unexported funcs, with lots of tests!

### PR DESCRIPTION
**Pull request by hasAdamr**
This pull replaces the code that generated `server_handler.go` and modified the functions `server_handler.go` to match the defined server methods in the definition file. 

Use:
```
// `service.proto`
package test
...
rpc Echo(In) returns (Out) {}
rpc Louder(In) returns (Out) {}
rpc Missing(In) returns (Out) {}
```
On:
```
package handler
...
func init() { /* Will keep unexported */ }
func (s wrong) Louder(..., in *pb.Wrong) (*pb.Wrong, ...) { /* Will remove per func recv */}
func (s testService) Echo (..., in *pb.Wrong) (*pb.Wrong, ...) { /* Will fix in and out params */ logic... }
func WhyIsThisFunctionExported() { /* Will remove pre being exported and not in the service defintion */ }
```
To get:

```
package handler
...
func init() { /* Will keep unexported */ }
func (s testService) Echo (..., in *pb.In) (*pb.Out, ...) { /* Will fix in and out params */ logic... }
func (s testServuce) Louder (..., in *pb.In) (*pb.Out, ...) { /* default */ ... }
func (s testServuce) Missing (..., in *pb.In) (*pb.Out, ...) { /* default */ ... }
```

This pull removes astmodifer, which was having panic problems, did not check the function receiver, did not update the in and out params, removed upexported functions (wanted at the time), and was untested.

Consequently fixes issue #59.

### CODE REVIEW

**Comment by hasLeland** [comment]
Package `go-truss/gengokit/generator` needs a package level description.

File: gengokit/gentesthelper/helper_test.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package gentesthelper
+
+import (
+	"testing"
+)
+
+func TestDiffGoCode(t *testing.T) {
+
```
Comment (hasLeland) [inline gengokit/gentesthelper/helper_test.go:32]: Is this meant to be an empty test?

File: gengokit/handler/handler.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
+// package handler parses service handlers and add/removes exported methods to
```
Comment (hasLeland) [inline gengokit/handler/handler.go]: May be more straightforward to write as:

```
Package handler manages the exported methods in the service handler code, adding/removing
exported methods to match the defined service methods.
```

File: gengokit/handler/handler.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+		return false
+	}
+
+	log.WithField("Func", name).
+		Debug("Method already exists in service definition; ignoring")
+
+	return true
+}
+
+// mRecvTypeString returns accepts and *ast.FuncDecl.Recv recv, and returns the
```
Comment (hasLeland) [inline gengokit/handler/handler.go]:
```
mRecvTypeString returns accepts and *ast.FuncDecl.Recv recv ...
```

I think this doc string needs to be re-written

File: gengokit/handler/handler.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+// string of the recv type.
+// func (s Foo) Test() {} -> "Foo"
```
Comment (hasLeland) [inline gengokit/handler/handler.go]: If this visual description is going to be provided, it should be indented to indicate it's not a continuation of the english description above.

File: gengokit/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+func mRecvTypeString(recv *ast.FieldList) string {
+	// func NoRecv {}
```
Comment (hasLeland) [inline gengokit/handler/handler.go]: Pretty sure the above is code which should be cleaned.

File: gengokit/handler/handler.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		return ""
+	}
+
+	return exprString(recv.List[0].Type)
+}
+
+// exprString returns the string representation of
+// ast.Expr for function receivers, parameters, and results.
+func exprString(e ast.Expr) string {
+	var hasPtr string
```
Comment (hasLeland) [inline gengokit/handler/handler.go]: `hasPtr` is named as though it's a bool, as in "does this have a pointer". However, it's being used to hold the prefix of the returned string. It's name should probably reflect that, maybe with `exprPrefix` or just `prefix`.

**Comment by hasLeland** [comment]
Besides what are pretty much style comments, looks like a welcome change!

File: gengokit/handler/handler.go (hunk 5 above)
Comment (hasAdamr) [reply gengokit/handler/handler.go]: That is an example of what the function does. I.e. the code directly below it handles the case of a function declaration with no receiver

```
func NoRecv() {}
```

Could be changed to

```
// handles functions with no receiver. `func NoRecv() {}
```

File: gengokit/gentesthelper/helper_test.go (hunk 1 above)
Comment (hasAdamr) [reply gengokit/gentesthelper/helper_test.go:32]: Well you can see there wan an intent to test it. Woops, I do wanna do a few cases here though.

File: gengokit/generator/gen.go [hunk 7]
```diff
@@ -70,32 +70,24 @@ func generateResponseFile(templFP string, te *gengokit.TemplateExecutor, prevGen
...
-	// Map of template paths to template rendering functions
-	renderTempl := map[string]string{
-		"NAME-service/handlers/server/server_handler.gotemplate": serverMethods,
-	}
-
-	// If we are rendering a template with a renderFunc and the file existed
-	// previously then use that function
-	if templ := renderTempl[templFP]; templ != "" {
+	// If we are rendering the server and or the client
+	if templFP == "NAME-service/handlers/server/server_handler.gotemplate" {
```
Comment (hasIan) [inline gengokit/generator/gen.go:75]: Using the constant `serverTemplPath` in this package will alleviate the need to use this "magic string." Single point of truth and all. (There's also a mention in the test, but that's less of a concern. Not that this is too much of a concern.)

File: gengokit/gentesthelper/helper_test.go [hunk 8]
```diff
@@ -1,9 +1,39 @@
...
 	"testing"
 )
 
 func TestDiffGoCode(t *testing.T) {
+	code := []string{
+		// Normal
+		`func Foo() {
+			println("test")
+		}`,
+		// Trailing whitespace on both ends
```
Comment (hasIan) [inline gengokit/gentesthelper/helper_test.go]: Nit: it's only trailing after the code. It's "surrounded blank lines," to be specific. Really really minor, but I did pause for a second trying to figure out what it meant before I realized it meant something different than it actually said.

File: gengokit/handler/handler.go [hunk 9]
```diff
@@ -241,11 +237,10 @@ func isValidFunc(f *ast.FuncDecl, m methodMap, pkgName string) bool {
 	return true
 }
 
-// mRecvTypeString returns accepts and *ast.FuncDecl.Recv recv, and returns the
+// mRecvTypeString accepts an *ast.FuncDecl.Recv recv, and returns the
```
Comment (hasIan) [inline gengokit/handler/handler.go]: Curious about the naming convention here. This looks like a throwback to C++ when some programmers used this convention to indicate that a variable was a class member. I'm not sure what it means here. Is there a good reason for it, or should it be renamed?

**Review by hasIan — COMMENTED** [review]
Minor comments that don't block a merge. Make whatever changes you think are appropriate. :shipit: :100:

**Comment by hasIan** [comment]
Oh, I said ship it, but there are some merge conflicts. But I imagine that's not surprising with all the PRs being merged.

File: gengokit/handler/handler.go (hunk 9 above)
Comment (hasAdamr) [reply gengokit/handler/handler.go]: Hmmm... I did not even know why it was called that till I looked though the blames.

Seems it stood for `methodRecvTypeString`, I will change it to `recvTypeToString`.

File: gengokit/gentesthelper/helper_test.go (hunk 8 above)
Comment (hasAdamr) [reply gengokit/gentesthelper/helper_test.go]: Will change to `// Leading and trailing whitespace`

File: gengokit/generator/gen.go (hunk 7 above)
Comment (hasAdamr) [reply gengokit/generator/gen.go:75]: In a upcoming pull request this will be fixed so I am leaving it for now.

**Pull request merged**